boot/bootutil/src/loader.c

```diff
@@ -1247,116 +1297,104 @@ boot_go(struct boot_rsp *rsp)
     boot_data.imgs[0].sectors = slot0_sectors;
     boot_data.imgs[1].sectors = slot1_sectors;
 
-    /* Open boot_data image areas for the duration of this call. */
-    for (slot = 0; slot < BOOT_NUM_SLOTS; slot++) {
-        fa_id = flash_area_id_from_image_slot(slot);
-        rc = flash_area_open(fa_id, &BOOT_IMG_AREA(&boot_data, slot));
+    /* The ARMv8-M architecture can contain more than one bootable
+     * image(secure and non-secure). The img_cnt paramter indicates how many
+     * image must be authenticated. Three flash partitions belongs to one
+     * bootable image:
+     *          - primary slot(image always executes from here)
+     *          - secondary slot(partitons for new image during sw upgrade)
+     *          - scratch area(used during image swapping)
+     * The flash_map_entry structure is initilized at the beginning of the
+     * loop to contain the addreses of the corresponding flash partitions,
+     * which belongs the current image.
+     */
+    for (img_idx = 0 ; img_idx < img_cnt; img_idx++) {
+        /* Initialize flash partitoin map */
+        flash_partition_map_init(img_idx);
+
+        /* Open boot_data image areas for the duration of this call. */
+        for (slot = 0; slot < BOOT_NUM_SLOTS; slot++) {
+            fa_id = flash_area_id_from_image_slot(slot);
+            rc = flash_area_open(fa_id, &BOOT_IMG_AREA(&boot_data, slot));
+            assert(rc == 0);
+        }
+        rc = flash_area_open(FLASH_AREA_IMAGE_SCRATCH,
+                             &BOOT_SCRATCH_AREA(&boot_data));
         assert(rc == 0);
-    }
-    rc = flash_area_open(FLASH_AREA_IMAGE_SCRATCH,
-                         &BOOT_SCRATCH_AREA(&boot_data));
-    assert(rc == 0);
-
-    /* Determine the sector layout of the image slots and scratch area. */
-    rc = boot_read_sectors();
-    if (rc != 0) {
-        goto out;
-    }
 
-    /* Attempt to read an image header from each slot. */
-    rc = boot_read_image_headers();
-    if (rc != 0) {
-        goto out;
-    }
+        /* Determine the sector layout of the image slots and scratch area. */
+        rc = boot_read_sectors();
+        if (rc != 0) {
+            goto out;
+        }
 
-    /* If the image slots aren't compatible, no swap is possible.  Just boot
-     * into slot 0.
-     */
-    if (boot_slots_compatible()) {
-        rc = boot_swap_if_needed(&swap_type);
-        assert(rc == 0);
+        /* Attempt to read an image header from each slot. */
+        rc = boot_read_image_headers();
         if (rc != 0) {
             goto out;
         }
 
-        /*
-         * The following states need image_ok be explicitly set after the
-         * swap was finished to avoid a new revert.
+        /* If the image slots aren't compatible, no swap is possible.  Just boot
```

Comment:
I'm not sure this makes sense in a multi-image context. We could install an upgrade in one "image", but fail to do so in another if the boot slots aren't compatible. This would lead to a partial upgrade, which I don't believe is what you want -- am I missing something? It seems like we need to make sure that success or failure to install upgrades should happen more uniformly across the image areas, or perhaps be made configurable (though fewer configuration knobs would be better IMO).

In general, I think the semantics around partial upgrade need to be specified clearly and the error handling behavior described.